`ERR_PNPM_OTHER_PM_EXPECTED  This project is configured to use yarn` while installing git-hosted package

### Last pnpm version that worked

v8.15.6

### pnpm version

v9.0.0-beta.2

### Code to reproduce the issue

`cd $(mktemp -d) && pnpm add github:isaacs/ts-node#isaacs/node-20-fix`

### Expected behavior

It works.

### Actual behavior

```
 WARN  Could not find preferred package ts-node@https://codeload.github.com/isaacs/ts-node/tar.gz/b614b1be4d3dacbaf26b31a9320e31287eaf4774(patch_hash=cstrxxonjryn63zbkj6xbnmhqa)(@swc/wasm@1.3.107)(@types/node@18.19.28)(typescript@5.3.3) in lockfile
../.local/share/pnpm/store/v3/tmp/_tmp_413332_0049e7542f9f5b8516166f53090f5745 [ts-node@10.9.1]: Running yarn-install script, done in 6.2s
../.local/share/pnpm/store/v3/tmp/_tmp_413332_0049e7542f9f5b8516166f53090f5745 [ts-node@10.9.1]: Running prepack script, failed in 952ms
...3332_0049e7542f9f5b8516166f53090f5745 prepack$ node ./scripts/prepack.mjs
│  ERR_PNPM_OTHER_PM_EXPECTED  This project is configured to use yarn
└─ Failed in 952ms at /root/.local/share/pnpm/store/v3/tmp/_tmp_413332_0049e7542f9f5b8516166f53090f5745
 ERR_PNPM_PREPARE_PACKAGE  Failed to prepare git-hosted package fetched from "https://codeload.github.com/isaacs/ts-node/tar.gz/b614b1be4d3dacbaf26b31a9320e31287eaf4774": ts-node@10.9.1 prepack: `node ./scripts/prepack.mjs`
Exit status 1

This error happened while installing a direct dependency of /root/an-website
Progress: resolved 29, reused 29, downloaded 0, added 0
```

### Additional information

_No response_

### Node.js version

v18.19.1

### Operating System

Linux

--- a/.changeset/friendly-tips-search.md
+++ b/.changeset/friendly-tips-search.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/prepare-package": patch
+"pnpm": patch
+---
+
+When building git-hosted dependencies, use the package manager required by the project [#7850](https://github.com/pnpm/pnpm/issues/7850).

--- a/exec/prepare-package/src/index.ts
+++ b/exec/prepare-package/src/index.ts
@@ -46,8 +46,15 @@ export async function preparePackage (opts: PreparePackageOptions, gitRootDir: s
     await runLifecycleHook(installScriptName, manifest, execOpts)
     for (const scriptName of PREPUBLISH_SCRIPTS) {
       if (manifest.scripts[scriptName] == null || manifest.scripts[scriptName] === '') continue
+      let newScriptName
+      if (pm !== 'pnpm') {
+        newScriptName = `${pm}-run-${scriptName}`
+        manifest.scripts[newScriptName] = `${pm} run ${scriptName}`
+      } else {
+        newScriptName = scriptName
+      }
       // eslint-disable-next-line no-await-in-loop
-      await runLifecycleHook(scriptName, manifest, execOpts)
+      await runLifecycleHook(newScriptName, manifest, execOpts)
     }
   } catch (err: unknown) {
     assert(util.types.isNativeError(err))

